File: TidbitsTrivia/Core/Models/GameMode.swift
import SwiftUI

/// The v1 single-player modes. Multiplayer modes (head-to-head, team,
/// living-room) layer on top in Phase 2 — they reuse the same GameEngine
/// loop, only the win condition and scoring shell differ.
enum GameMode: String, CaseIterable, Identifiable, Sendable {
    case classic     // 10 questions, accuracy + speed bonus
    case timeAttack  // as many as you can in 60s
    case survival    // keep going until one wrong answer
    case stake       // bet a fixed budget of confidence chips per question
    case sweep       // fill a themed set; beat your own best (count-scored)
    case pictureId   // identify the subject from its image (E1 enrichment)
    case thisOrThat  // binary pick — which came first (E1 chronology)
    case closestCall // estimate a number on a slider, scored by proximity (E1)
    case ordering    // arrange items chronologically, partial credit (E1)
    case matching    // link each key to its value, partial credit (E1)
    case typeAnswer  // free-text recall, alias-matched (E1); tvOS self-marks
    case oddOneOut   // which doesn't belong — plain MCQ, outlier is the answer
    case daily       // one fixed daily set, streak-bearing, shareable

    var id: String { rawValue }

    var title: String {
        switch self {
        case .classic:    return "Classic"
        case .timeAttack: return "Time Attack"
        case .survival:   return "Survival"
        case .stake:      return "Stake"
        case .sweep:      return "Sweep"
        case .pictureId:  return "Picture ID"
        case .thisOrThat: return "Which First?"
        case .closestCall: return "Closest Call"
        case .ordering:   return "In Order"
        case .matching:   return "Match Up"
        case .typeAnswer: return "Name It"
        case .oddOneOut:  return "Odd One Out"
        case .daily:      return "Daily Tidbit"
        }
    }

    var blurb: String {
        switch self {
        case .classic:    return "Ten questions. Speed counts."
        case .timeAttack: return "How many in 60 seconds?"
        case .survival:   return "One wrong answer ends it."
        case .stake:      return "Bet your confidence. No risk."
        case .sweep:      return "Fill the set. Beat your best."
        case .pictureId:  return "Name what you see."
        case .thisOrThat: return "Which came first?"
        case .closestCall: return "How close can you get?"
        case .ordering:   return "Arrange them in time."
        case .matching:   return "Link each pair."
        case .typeAnswer: return "Type the answer."
        case .oddOneOut:  return "Which doesn't belong?"
        case .daily:      return "Everyone's puzzle. Keep your streak."
        }
    }

    var symbol: String {
        switch self {
        case .classic:    return "list.number"
        case .timeAttack: return "timer"
        case .survival:   return "heart.fill"
        case .stake:      return "chart.bar.fill"
        case .sweep:      return "square.grid.3x3.fill"
        case .pictureId:  return "photo.fill"
        case .thisOrThat: return "arrow.left.arrow.right"
        case .closestCall: return "target"
        case .ordering:   return "arrow.up.arrow.down"
        case .matching:   return "link"
        case .typeAnswer: return "keyboard"
        case .oddOneOut:  return "questionmark.diamond.fill"
        case .daily:      return "sun.max.fill"
        }
    }

    var accent: Color {
        switch self {
        case .classic:    return Tidbits.Palette.blue
        case .timeAttack: return Tidbits.Palette.coral
        case .survival:   return Tidbits.Palette.grape
        case .stake:      return Tidbits.Palette.mint
        case .sweep:      return Tidbits.Palette.teal
        case .pictureId:  return Tidbits.Palette.pink
        case .thisOrThat: return Tidbits.Palette.grape
        case .closestCall: return Tidbits.Palette.yellow
        case .ordering:   return Tidbits.Palette.blue
        case .matching:   return Tidbits.Palette.coral
        case .typeAnswer: return Tidbits.Palette.mint
        case .oddOneOut:  return Tidbits.Palette.grape
        case .daily:      return Tidbits.Palette.yellow
        }
    }

    /// Per-question time budget in seconds (nil = the mode's own clock).
    var perQuestionSeconds: Double? {
        switch self {
        case .classic:    return 20
        case .timeAttack: return nil   // global 60s clock
        case .survival:   return 15
        case .stake:      return 30    // generous — calibration shouldn't be rushed
        case .sweep:      return 12    // rapid-fire, but a clock that never punishes score
        case .pictureId:  return 20    // time to look at the image
        case .thisOrThat: return 12    // snap binary call
        case .closestCall: return 25   // estimation deserves a beat of thought
        case .ordering:   return 35    // reordering takes a moment
        case .matching:   return 40    // linking pairs takes thought
        case .typeAnswer: return 25
        case .oddOneOut:  return 20
        case .daily:      return 30
        }
    }

    var questionCount: Int {
        switch self {
        case .classic:    return 10
        case .timeAttack: return 99    // bounded by the clock
        case .survival:   return 99    // bounded by a wrong answer
        case .stake:      return 8     // matches the confidence-chip budget
        case .sweep:      return 12    // a "set" to fill — the grid is the scoreboard
        case .pictureId:  return 10
        case .thisOrThat: return 10
        case .closestCall: return 8
        case .ordering:   return 6
        case .matching:   return 6
        case .typeAnswer: return 8
        case .oddOneOut:  return 8
        case .daily:      return 7
        }
    }

    /// Stake mode only: the fixed budget of confidence chips for one round
    /// (sum of `count` == questionCount). Spending more on one question means
    /// fewer chips for the rest — that scarcity is what makes it calibration,
    /// not "stake max on everything." Adds-only: a wrong answer earns 0 but the
    /// chip is spent; the score can never go negative (Decision 022).
    static let stakeBudget: [(value: Int, label: String, count: Int)] = [
        (3, "Sure", 2), (2, "Likely", 3), (1, "Hunch", 3),
    ]

    var globalClockSeconds: Double? { self == .timeAttack ? 60 : nil }
}

File: windows/Tidbits.Core/Models/GameMode.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace Tidbits.Core.Models;

/// The game modes (port of Core/Models/GameMode.swift). The string IDs are the
/// wire values — keep them byte-identical to the Swift rawValues. Serializes as
/// that string (not the enum ordinal) for cross-platform wire compatibility.
[JsonConverter(typeof(GameModeJsonConverter))]
public enum GameMode
{
    Classic, TimeAttack, Survival, Stake, Sweep, PictureId, ThisOrThat, ClosestCall,
    Ordering, Matching, TypeAnswer, OddOneOut, Ladder, Enumerate, BarTrivia, Mix, Daily,
}

/// (De)serializes GameMode as its wire string ("classic", "timeAttack", …).
public sealed class GameModeJsonConverter : JsonConverter<GameMode>
{
    public override GameMode Read(ref Utf8JsonReader reader, Type _, JsonSerializerOptions __) =>
        GameModeExtensions.FromId(reader.GetString() ?? "") ?? GameMode.Classic;

    public override void Write(Utf8JsonWriter writer, GameMode value, JsonSerializerOptions _) =>
        writer.WriteStringValue(value.Id());
}

public readonly record struct StakeChip(int Value, string Label, int Count);

public static class GameModeExtensions
{
    /// The wire ID (Swift rawValue) — used in RTDB/JSON/persistence.
    public static string Id(this GameMode m) => m switch
    {
        GameMode.Classic => "classic",
        GameMode.TimeAttack => "timeAttack",
        GameMode.Survival => "survival",
        GameMode.Stake => "stake",
        GameMode.Sweep => "sweep",
        GameMode.PictureId => "pictureId",
        GameMode.ThisOrThat => "thisOrThat",
        GameMode.ClosestCall => "closestCall",
        GameMode.Ordering => "ordering",
        GameMode.Matching => "matching",
        GameMode.TypeAnswer => "typeAnswer",
        GameMode.OddOneOut => "oddOneOut",
        GameMode.Ladder => "ladder",
        GameMode.Enumerate => "enumerate",
        GameMode.BarTrivia => "barTrivia",
        GameMode.Mix => "mix",
        GameMode.Daily => "daily",
        _ => "classic",
    };

    public static GameMode? FromId(string id) =>
        System.Enum.GetValues<GameMode>().Cast<GameMode?>().FirstOrDefault(m => m!.Value.Id() == id);

    public static string Title(this GameMode m) => m switch
    {
        GameMode.Classic => "Classic",
        GameMode.TimeAttack => "Time Attack",
        GameMode.Survival => "Survival",
        GameMode.Stake => "Stake",
        GameMode.Sweep => "Sweep",
        GameMode.PictureId => "Picture ID",
        GameMode.ThisOrThat => "Which First?",
        GameMode.ClosestCall => "Closest Call",
        GameMode.Ordering => "In Order",
        GameMode.Matching => "Match Up",
        GameMode.TypeAnswer => "Name It",
        GameMode.OddOneOut => "Odd One Out",
        GameMode.Ladder => "Ladder",
        GameMode.Enumerate => "Name as Many",
        GameMode.BarTrivia => "Trivia Night",
        GameMode.Mix => "Custom Mix",
        GameMode.Daily => "Daily Tidbit",
        _ => "Classic",
    };

    public static string Blurb(this GameMode m) => m switch
    {
        GameMode.Classic => "Ten questions. Speed counts.",
        GameMode.TimeAttack => "How many in 60 seconds?",
        GameMode.Survival => "One wrong answer ends it.",
        GameMode.Stake => "Bet your confidence. No risk.",
        GameMode.Sweep => "Fill the set. Beat your best.",
        GameMode.PictureId => "Name what you see.",
        GameMode.ThisOrThat => "Which came first?",
        GameMode.ClosestCall => "How close can you get?",
        GameMode.Ordering => "Arrange them in time.",
        GameMode.Matching => "Link each pair.",
        GameMode.TypeAnswer => "Type the answer.",
        GameMode.OddOneOut => "Which doesn't belong?",
        GameMode.Ladder => "Climb from easy to hard.",
        GameMode.Enumerate => "How many can you name?",
        GameMode.BarTrivia => "Host a night. Every kind of round.",
        GameMode.Mix => "Your picked modes, shuffled together.",
        GameMode.Daily => "Everyone's puzzle. Keep your streak.",
        _ => "",
    };

    /// The round title when used as a Trivia Night round type.
    public static string NightRoundTitle(this GameMode m) => m switch
    {
        GameMode.Classic => "General Knowledge",
        GameMode.PictureId => "Picture Round",
        GameMode.ThisOrThat => "Which Came First?",
        GameMode.ClosestCall => "Closest Wins",
        GameMode.Ordering => "Put Them In Order",
        GameMode.Matching => "Match-Up",
        GameMode.TypeAnswer => "Name It",
        GameMode.OddOneOut => "Odd One Out",
        GameMode.Enumerate => "Name As Many",
        _ => m.Title(),
    };

    /// Per-question time budget in seconds (null = the mode's own clock).
    public static double? PerQuestionSeconds(this GameMode m) => m switch
    {
        GameMode.Classic => 20,
        GameMode.TimeAttack => null, // global 60s clock
        GameMode.Survival => 15,
        GameMode.Stake => 30,
        GameMode.Sweep => 12,
        GameMode.PictureId => 20,
        GameMode.ThisOrThat => 12,
        GameMode.ClosestCall => 25,
        GameMode.Ordering => 35,
        GameMode.Matching => 40,
        GameMode.TypeAnswer => 25,
        GameMode.OddOneOut => 20,
        GameMode.Ladder => 20,
        GameMode.Enumerate => 60,
        GameMode.BarTrivia => 20,
        GameMode.Mix => 20,
        GameMode.Daily => 30,
        _ => 20,
    };

    public static int QuestionCount(this GameMode m) => m switch
    {
        GameMode.Classic => 10,
        GameMode.TimeAttack => 99,
        GameMode.Survival => 99,
        GameMode.Stake => 8,
        GameMode.Sweep => 12,
        GameMode.PictureId => 10,
        GameMode.ThisOrThat => 10,
        GameMode.ClosestCall => 8,
        GameMode.Ordering => 6,
        GameMode.Matching => 6,
        GameMode.TypeAnswer => 8,
        GameMode.OddOneOut => 8,
        GameMode.Ladder => 10,
        GameMode.Enumerate => 3,
        GameMode.BarTrivia => 20,
        GameMode.Mix => 10,
        GameMode.Daily => 7,
        _ => 10,
    };

    public static double? GlobalClockSeconds(this GameMode m) => m == GameMode.TimeAttack ? 60 : null;

    /// Whether spaced-review misses (plain corpus MCQ) may be woven into a round.
    public static bool AcceptsReview(this GameMode m) => m switch
    {
        GameMode.Classic or GameMode.TimeAttack or GameMode.Survival
            or GameMode.Stake or GameMode.Sweep or GameMode.Ladder => true,
        _ => false,
    };

    /// Stake mode: the fixed budget of confidence chips for one round
    /// (sum of Count == QuestionCount). Adds-only — a wrong answer earns 0.
    public static readonly IReadOnlyList<StakeChip> StakeBudget = new[]
    {
        new StakeChip(3, "Sure", 2),
        new StakeChip(2, "Likely", 3),
        new StakeChip(1, "Hunch", 3),
    };
}
